src/kernel/arch/x86/cmos.zig

```diff
@@ -0,0 +1,458 @@
+
+/// The register location for returning the day, (0 - 31).
+const REGISTER_DAY: u8 = 0x07;
+
+/// The register location for returning the month, (0 - 11).
+const REGISTER_MONTH: u8 = 0x08;
+
+/// The register location for returning the year, (0 - 99).
+const REGISTER_YEAR: u8 = 0x09;
+
+/// The register location for returning the century.
+const REGISTER_CENTURY: u8 = 0x32;
+
+/// The register location for return the status A register.
+const STATUS_REGISTER_A: u8 = 0x0A;
+
+/// The register location for return the status B register.
+const STATUS_REGISTER_B: u8 = 0x0B;
+
+/// The register location for return the status C register.
+const STATUS_REGISTER_C: u8 = 0x0C;
+
+/// The non-mockable interrupts are on the 8th bit of port 0x70 which is the register select port
+/// for the CMOS. This will need to be disabled when selecting CMOS registers.
+const NMI_BIT: u8 = 0x80;
+
+/// The enum for selecting the status register to read from
+pub const StatusRegister = enum {
+    /// Status register A
+    A,
+
+    /// Status register B
+    B,
+
+    /// Status register C
+    C,
+};
+
+/// The enum for selecting the real time clock registers.
+pub const RtcRegister = enum {
+    /// The seconds register
+    SECOND,
+
+    /// The minutes register
+    MINUTE,
+
+    /// The hours register
+    HOUR,
+
+    /// The days register
+    DAY,
+
+    /// The months register
+    MONTH,
+
+    /// The year register
+    YEAR,
+
+    /// The century register
+    CENTURY,
+};
+
+///
+/// Tell the CMOS chip to select the given register ready for read or writing to.
+///
+/// Arguments:
+///     IN reg: u8 - The register index to select in the CMOS chip.
+///
+inline fn selectRegister(reg: u8) void {
+    arch.outb(ADDRESS, reg);
+}
+
+///
+/// Tell the CMOS chip to select the given register ready for read or writing to. This also
+/// disables the NMI.
+///
+/// Arguments:
+///     IN reg: u8 - The register index to select in the CMOS chip.
+///
+inline fn selectRegisterNoNMI(reg: u8) void {
+    arch.outb(ADDRESS, reg | NMI_BIT);
+}
+
+///
+/// Write to the selected register to the CMOS chip.
+///
+/// Arguments:
+///     IN data: u8 - The data to write to the selected register.
+///
+inline fn writeRegister(data: u8) void {
+    arch.outb(DATA, data);
+}
+
+///
+/// Read the selected register from the CMOS chip.
+///
+/// Return: u8
+///     The value in the selected register.
+///
+inline fn readRegister() u8 {
+    return arch.inb(DATA);
+}
+
+///
+/// Select then read a register from the CMOS chip. This include a I/O wait to ensure the CMOS chip
+/// has time to select the register.
+///
+/// Arguments:
+///     IN reg: u8                    - The register index to select in the CMOS chip.
+///     IN comptime disable_nmi: bool - Whether to disable NMI when selecting a register.
+///
+/// Return: u8
+///     The value in the selected register.
+///
+inline fn readSelectedRegister(reg: u8, comptime disable_nmi: bool) u8 {
```

Comment:
I think a better API would be to take a StatusRegister instance. Also, `readSelectedRegister` sounds as though it reads an already-selected register. Perhaps `selectAndReadRegister` would be better?